sdk/src/androidTest/java/com/bugsnag/android/ConcurrentCallbackTest.java

```diff
@@ -0,0 +1,74 @@
+package com.bugsnag.android;
+
+import android.support.annotation.NonNull;
+import android.support.test.filters.SmallTest;
+import android.support.test.runner.AndroidJUnit4;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.Collection;
+
+/**
+ * Ensures that if a callback is added or removed during iteration, a
+ * {@link java.util.ConcurrentModificationException} is not thrown
+ */
+@RunWith(AndroidJUnit4.class)
+@SmallTest
+public class ConcurrentCallbackTest {
+
+    private Client client;
+
+    @Before
+    public void setUp() throws Exception {
+        client = BugsnagTestUtils.generateClient();
+    }
+
+    @Test
+    public void testClientNotifyModification() throws Exception {
+        final Collection<BeforeNotify> beforeNotifyTasks = client.config.getBeforeNotifyTasks();
+        client.beforeNotify(new BeforeNotify() {
+            @Override
+            public boolean run(Error error) {
+                beforeNotifyTasks.add(new BeforeNotifySkeleton());
+                // modify the Set, when iterating to the next callback this should not crash
+                return true;
```

Comment:
This test fails without the change?

Reply:
Yes, as it would modify the contents of the `LinkedHashSet` when an iterator has already been implicitly created in [`runBeforeNotifyTasks`](https://github.com/bugsnag/bugsnag-android/blob/master/sdk/src/main/java/com/bugsnag/android/Client.java#L1274)